From 8b627f96728e85894626fe3f7657cd5968ca9b07 Mon Sep 17 00:00:00 2001
From: Simon McVittie <smcv@debian.org>
Date: Wed, 13 Apr 2016 07:14:59 +0100
Subject: Add a systemd user service for the terminal server

This is used to implement the D-Bus session service if
"dbus-daemon --session" was configured to use systemd activation,
most commonly via the units installed by configuring dbus 1.10+
with --enable-user-session.

With this service, child processes such as the user's shell
end up in the cgroup .../user@UID.service/org.gnome.Terminal.service
instead of in .../user@UID.service/dbus.service.

Bug: https://bugzilla.gnome.org/show_bug.cgi?id=759115
Applied-upstream: 3.21.0, commit:4f34660f0d1e7a61eaed6123d314a69664a7adeb
---
 configure.ac    |  1 +
 src/Makefile.am | 17 +++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/configure.ac b/configure.ac
index 513cc47..701723d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -25,6 +25,7 @@ fi
 AM_MAINTAINER_MODE([enable])
 
 PKG_PROG_PKG_CONFIG
+AC_PROG_MKDIR_P
 AC_PROG_CC
 AC_ISC_POSIX
 AC_HEADER_STDC
diff --git a/src/Makefile.am b/src/Makefile.am
index 17f55e9..e17848e 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -16,6 +16,11 @@ endif # WITH_NAUTILUS_EXTENSION
 
 dbusservice_DATA = org.gnome.Terminal.service
 
+# It is correct for this to be in ${prefix}/lib, even on systems where that
+# does not match ${libdir}. This is what systemd uses on such platforms.
+systemd_userdir = $(prefix)/lib/systemd/user
+systemd_user_DATA = gnome-terminal-server.service
+
 gnome_terminal_server_built = \
 	terminal-marshal.c \
 	terminal-marshal.h \
@@ -140,8 +145,19 @@ terminal-marshal.c: $(srcdir)/terminal-marshal.list
 org.gnome.Terminal.service: Makefile
 	$(AM_V_GEN) ( echo "[D-BUS Service]"; \
 		      echo "Name=org.gnome.Terminal"; \
+		      echo "SystemdService=gnome-terminal-server.service"; \
 		      echo "Exec=${libexecdir}/gnome-terminal-server") > $@
 
+gnome-terminal-server.service: Makefile
+	$(AM_V_GEN) ( echo "[Unit]"; \
+		      echo "Description=GNOME Terminal"; \
+		      echo "[Service]"; \
+		      echo "# Don't kill screen sessions, backgrounded X apps etc."; \
+		      echo "KillMode=process"; \
+		      echo "Type=dbus"; \
+		      echo "BusName=org.gnome.Terminal"; \
+		      echo "ExecStart=${libexecdir}/gnome-terminal-server") > $@
+
 terminal-gdbus-generated.c terminal-gdbus-generated.h: org.gnome.Terminal.xml Makefile
 	$(AM_V_GEN) $(GDBUS_CODEGEN) \
 			--interface-prefix=org.gnome.Terminal \
@@ -396,6 +412,7 @@ CLEANFILES = \
 	gnome-terminal.schemas \
 	stamp-terminal-type-builtins.h \
 	org.gnome.Terminal.service \
+	gnome-terminal-server.service \
 	$(BUILT_SOURCES)
 
 EXTRA_DIST = \
